libdnf/sack/query.cpp

```diff
@@ -43,6 +43,85 @@
 
 namespace libdnf {
 
+struct NevraID {
+    Id name;
+    Id arch;
+    Id evr;
+    /**
+    * @brief Parsing function for nevra string into name, evr, arch and transorming it into libsolv
+    * Id
+    *
+    * @return bool Returns true if parsing succesfull and all elements is known to pool
+    */
+    bool parse(Pool * pool, const char * nevraPattern);
+};
+
+bool
+NevraID::parse(Pool * pool, const char * nevraPattern)
+{
+    const char * evrDelim = nullptr;
+    const char * releaseDelim = nullptr;
+    const char * archDelim = nullptr;
+    const char * end;
+
+    // parse nevra
+    for (end = nevraPattern; *end != '\0'; ++end) {
+        if (*end == '-') {
+            evrDelim = releaseDelim;
+            releaseDelim = end;
+        } else if (*end == '.') {
+            archDelim = end;
+        }
+    }
+
+    // test name presence
+    if (!evrDelim || evrDelim == nevraPattern)
+        return false;
+
+    auto nameLen = evrDelim - nevraPattern;
+
+    // strip epoch "0:"
```

Comment:
you actually don't need this ;) libsolv will work fine with it too I think

Reply:
I am sorry, but you need it if you don't want to use evr compare function but only id to id comparison.